In [6]:
df <- read.csv("E:/Machine Learning/Datasets/Classification+preprocessed+data+R.csv", header = TRUE)
dim(df)

[1] 506  17

In [7]:
View(df)

price resid_area air_qual room_num age   teachers poor_prop airportYES
1   24.0  32.31      0.538    6.575     65.2 24.7      4.98     1         
2   21.6  37.07      0.469    6.421     78.9 22.2      9.14     0         
3   34.7  37.07      0.469    7.185     61.1 22.2      4.03     0         
4   33.4  32.18      0.458    6.998     45.8 21.3      2.94     1         
5   36.2  32.18      0.458    7.147     54.2 21.3      5.33     0         
6   28.7  32.18      0.458    6.430     58.7 21.3      5.21     1         
7   22.9  37.87      0.524    6.012     66.6 24.8     12.43     1         
8   22.1  37.87      0.524    6.172     96.1 24.8     19.15     0         
9   16.5  37.87      0.524    5.631    100.0 24.8     29.93     1         
10  18.9  37.87      0.524    6.004     85.9 24.8     17.10     1         
11  15.0  37.87      0.524    6.377     94.3 24.8     20.45     0         
12  18.9  37.87      0.524    6.009     82.9 24.8     13.27     0         
13  21.7  37.87      0.524    5.889     39.0 24.8     15.71     1         
14  20.4  38.14      0.538    5.949     61.8 19.0      8.26     1         
15  18.2  38.14      0.538    6.096     84.5 19.0     10.26     0         
16  19.9  38.14      0.538    5.834     56.5 19.0      8.47     1         
17  23.1  38.14      0.538    5.935     29.3 19.0      6.58     0         
18  17.5  38.14      0.538    5.990     81.7 19.0     14.67     0         
19  20.2  38.14      0.538    5.456     36.6 19.0     11.69     1         
20  18.2  38.14      0.538    5.727     69.5 19.0     11.28     0         
21  13.6  38.14      0.538    5.570     98.1 19.0     21.02     1         
22  19.6  38.14      0.538    5.965     89.2 19.0     13.83     1         
23  15.2  38.14      0.538    6.142     91.7 19.0     18.72     1         
24  14.5  38.14      0.538    5.813    100.0 19.0     19.88     1         
25  15.6  38.14      0.538    5.924     94.1 19.0     16.30     1         
26  13.9  38.14      0.538    5.599     85.7 19.0     16.51     1         
27  16.6  38.14      0.538    5.813     90.3 19.0     14.81     0         
28  14.8  38.14      0.538    6.047     88.8 19.0     17.28     1         
29  18.4  38.14      0.538    6.495     94.4 19.0     12.80     1         
30  21.0  38.14      0.538    6.674     87.3 19.0     11.98     0         
⋮   ⋮     ⋮          ⋮        ⋮        ⋮     ⋮        ⋮         ⋮         
477 16.7  48.10      0.614    6.484    93.6  19.8     18.68     1         
478 12.0  48.10      0.614    5.304    97.3  19.8     24.91     0         
479 14.6  48.10      0.614    6.185    96.7  19.8     18.03     1         
480 21.4  48.10      0.614    6.229    88.0  19.8     13.11     0         
481 23.0  48.10      0.532    6.242    64.7  19.8     10.74     0         
482 23.7  48.10      0.532    6.750    74.9  19.8      7.74     0         
483 25.0  48.10      0.532    7.061    77.0  19.8      7.01     0         
484 21.8  48.10      0.532    5.762    40.3  19.8     10.42     0         
485 20.6  48.10      0.583    5.871    41.9  19.8     13.34     1         
486 21.2  48.10      0.583    6.312    51.9  19.8     10.58     0         
487 19.1  48.10      0.583    6.114    79.8  19.8     14.98     1         
488 20.6  48.10      0.583    5.905    53.2  19.8     11.45     0         
489 15.2  57.74      0.609    5.454    92.7  19.9     18.06     0         
490  7.0  57.74      0.609    5.414    98.3  19.9     23.97     0         
491  8.1  57.74      0.609    5.093    98.0  19.9     29.68     1         
492 13.6  57.74      0.609    5.983    98.8  19.9     18.07     0         
493 20.1  57.74      0.609    5.983    83.5  19.9     13.35     1         
494 21.8  39.69      0.585    5.707    54.0  20.8     12.01     1         
495 24.5  39.69      0.585    5.926    42.6  20.8     13.59     1         
496 23.1  39.69      0.585    5.670    28.8  20.8     17.60     0         
497 19.7  39.69      0.585    5.390    72.9  20.8     21.14     0         
498 18.3  39.69      0.585   

## Predicting the overall accuracy

In [8]:
glm.fit <- glm(Sold~., data=df, family = binomial)
summary(glm.fit)

glm.probs <- predict(glm.fit, type = "response")
glm.probs[1:10]

glm.pred <- rep("NO", 506)
glm.pred[glm.probs>0.5] = "YES"

table(glm.pred, df$Sold)


Call:
glm(formula = Sold ~ ., family = binomial, data = df)

Coefficients:
                         Estimate Std. Error z value Pr(>|z|)    
(Intercept)             -2.138280   2.649317  -0.807  0.41961    
price                   -0.274084   0.032972  -8.313  < 2e-16 ***
resid_area               0.019211   0.026674   0.720  0.47140    
air_qual                -7.418344   2.691109  -2.757  0.00584 ** 
room_num                 1.106728   0.276645   4.001 6.32e-05 ***
age                     -0.001993   0.006595  -0.302  0.76247    
teachers                 0.315004   0.063810   4.937 7.95e-07 ***
poor_prop               -0.207690   0.033776  -6.149 7.80e-10 ***
airportYES              -0.152400   0.210677  -0.723  0.46944    
n_hos_beds               0.175963   0.071336   2.467  0.01364 *  
n_hot_rooms             -0.074215   0.051574  -1.439  0.15015    
waterbodyLake           -0.103179   0.303632  -0.340  0.73400    
waterbodyLake.and.River -0.062613   0.327945  -0.191  0.84858    


1          2          3          4          5          6          7 
0.85141600 0.67065484 0.03402370 0.27038629 0.09185864 0.20166690 0.43390218 
         8          9         10 
0.14510166 0.03966122 0.41134502

        
glm.pred   0   1
     NO  197  81
     YES  79 149

In [9]:
346/506*100

[1] 68.37945

# Test-Train Split

In [12]:
library(caTools)
set.seed(0)
split <- sample.split(df, SplitRatio = 0.8)
train <- subset(df, split == TRUE)
test <- subset(df, split == FALSE)

In [13]:
dim(train)
dim(test)

[1] 386  17

[1] 120  17

# Logistic Regression

In [14]:
train.fit <- glm(Sold~., data=train_set, family=binomial)
test.probs <- predict(train.fit, test_set, type='response')
test.pred <- rep('NO', 120)
test.pred[(test.probs > .5)] = 'YES'
table(test.pred, test_set$Sold)

         
test.pred  0  1
      NO  42 16
      YES 26 36

In [15]:
78/120*100

[1] 65

# Linear Discriminant Analysis

In [18]:
library(MASS)
trainLDA <- lda(Sold~., data=train)
test.predict <- predict(trainLDA, test)
test.predict$posterior
test.class <- test.predict$class
table(test.class, test$Sold)

0          1         
1   0.2071420  0.7928580 
5   0.8933769  0.1066231 
7   0.5928471  0.4071529 
8   0.8752496  0.1247504 
18  0.7597365  0.2402635 
22  0.6144932  0.3855068 
24  0.7577050  0.2422950 
25  0.6553961  0.3446039 
35  0.5048348  0.4951652 
39  0.5240466  0.4759534 
41  0.6984209  0.3015791 
42  0.4017050  0.5982950 
52  0.4526438  0.5473562 
56  0.8856250  0.1143750 
58  0.7662167  0.2337833 
59  0.7400102  0.2599898 
69  0.5584185  0.4415815 
73  0.3428322  0.6571678 
75  0.2520572  0.7479428 
76  0.3001061  0.6998939 
86  0.3118537  0.6881463 
90  0.3481278  0.6518722 
92  0.2025094  0.7974906 
93  0.1982796  0.8017204 
103 0.3830133  0.6169867 
107 0.6755323  0.3244677 
109 0.3638291  0.6361709 
110 0.4362251  0.5637749 
120 0.2396014  0.7603986 
124 0.6382037  0.3617963 
⋮   ⋮          ⋮         
381 0.04229004 0.95770996
382 0.24688002 0.75311998
392 0.79988606 0.20011394
396 0.32363359 0.67636641
398 0.17907492 0.82092508
399 0.51705841 0.48294159
409 0.85305791 0.14694209
413 0.98610402 0.01389598
415 0.94819525 0.05180475
416 0.38859657 0.61140343
426 0.54478995 0.45521005
430 0.27829670 0.72170330
432 0.27718739 0.72281261
433 0.09750721 0.90249279
443 0.39259800 0.60740200
447 0.33909021 0.66090979
449 0.35986654 0.64013346
450 0.25905975 0.74094025
460 0.73531703 0.26468297
464 0.28446358 0.71553642
466 0.52376029 0.47623971
467 0.59800928 0.40199072
477 0.45370021 0.54629979
481 0.41945738 0.58054262
483 0.27946958 0.72053042
484 0.61448077 0.38551923
494 0.62557571 0.37442429
498 0.67994917 0.32005083
500 0.45930746 0.54069254
501 0.52856402 0.47143598

          
test.class  0  1
         0 44 16
         1 24 36

In [19]:
80/120*100

[1] 66.66667

# K-Nearest Neighbors

In [20]:
library(class)
trainX <- train[,-16]
testX <- test[,-16]
trainY <- train$Sold
testY <- test$Sold
trainX_s <- scale(trainX)
testX_s <- scale(testX)
set.seed(0)

In [21]:
k=3
knn.pred <- knn(trainX_s, testX_s, trainY, k=k)
table(knn.pred,testY)

        testY
knn.pred  0  1
       0 38 24
       1 30 28

In [22]:
66/120*100

[1] 55

In [23]:
k=5
knn.pred <- knn(trainX_s, testX_s, trainY, k=k)
table(knn.pred,testY)

        testY
knn.pred  0  1
       0 42 27
       1 26 25

In [24]:
67/120*100

[1] 55.83333

# Conclusion

The LDA model has the highest accuracy on the test set as 66.67% and its values are the best for predicting whether a house will be sold within 3 months or not